### DISCUSSION TITLE
Visual Studio 2022 uses Ccache

### DISCUSSION THREAD

**Discussion by Elijah-777** [ccache/ccache#1148] [Support]
### Description ###

Visual Sdudio 2022 uses Ccache, like on MAC, to pass variables directly to CMake
set(CMAKE_C_COMPILER_LAUNCHER   "ccache")
Set (CMAKE_CXX_COMPILER_LAUNCHER ccache ")"
Or how to use ccache directly on VS now? Every time I use it, it doesn't work properly.

**Comment by cristianadam** [comment]
If you look at the [\<LANG\>_COMPILER_LAUNCHER](https://cmake.org/cmake/help/latest/prop_tgt/LANG_COMPILER_LAUNCHER.html) CMake documentation you  can see this part:

> The [Makefile Generators](https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html#makefile-generators) and the [Ninja](https://cmake.org/cmake/help/latest/generator/Ninja.html#generator:Ninja) generator will run this tool and pass the compiler and its arguments to the tool. Some example tools are distcc and ccache.

The "Visual Studio" generator is not supported by CMake, you will have to use the Ninja generator with Visual Studio in order to get "ccache" working as you expect.

**Comment by Elijah-777** [comment]
VS defaults to use ninja, and I set "set_property(GLOBAL PROPERTY RULE_LAUNCH_COMPILE "ccache") "in the Cmake file of my project, but it still doesn't work  ,thank you @cristianadam

**Comment by cristianadam** [comment]
`RULE_LAUNCH_COMPILE` should not be used by projects, in the [documentation](https://cmake.org/cmake/help/latest/prop_tgt/RULE_LAUNCH_COMPILE.html) you can see:

> Note This property is intended for internal use by [ctest(1)](https://cmake.org/cmake/help/latest/manual/ctest.1.html#manual:ctest(1)). Projects and developers should use the [\<LANG\>_COMPILER_LAUNCHER](https://cmake.org/cmake/help/latest/prop_tgt/LANG_COMPILER_LAUNCHER.html#prop_tgt:%3CLANG%3E_COMPILER_LAUNCHER) target properties or the associated [CMAKE_\<LANG\>_COMPILER_LAUNCHER](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_COMPILER_LAUNCHER.html#variable:CMAKE_%3CLANG%3E_COMPILER_LAUNCHER) variables instead.

Make sure you have `ccache` in path, or use `find_program` to have the full path to ccache.

I use `ccache` with the Visual C++ compiler and Qt Creator as IDE all the time. I have no idea how Visual Studio works using Ninja, I never tried getting it to work in that scenario.

**Comment by Elijah-777** [comment]
I'm going to set in CMakeList (CMAKE_C_COMPILER_LAUNCHER "ccache")
Set (CMAKE_CXX_COMPILER_LAUNCHER "ccache ") It still doesn't work @cristianadam

**Comment by Elijah-777** [comment]
![image](https://user-images.githubusercontent.com/98135989/186881982-8a735eb0-76e3-410d-be58-a071f4e2f3fe.png)
